backends/ent/backend.go

```diff
@@ -0,0 +1,90 @@
+// --------------------------------------------------------------
+// SPDX-FileCopyrightText: Copyright © 2024 The Protobom Authors
+// SPDX-FileType: SOURCE
+// SPDX-License-Identifier: Apache-2.0
+// --------------------------------------------------------------
+package ent
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"slices"
+
+	sqlite "github.com/glebarez/go-sqlite"
+	"github.com/protobom/protobom/pkg/storage"
+
+	"github.com/protobom/storage/internal/backends/ent"
+)
+
+// Backend implements the protobom.pkg.storage.Backend interface.
+type Backend struct {
+	// BackendOptions is the set of options common to all ent Backends.
+	*BackendOptions
+}
+
+var _ storage.Backend = (*Backend)(nil)
+
+func NewBackend(opts ...Option) *Backend {
+	backend := &Backend{
+		BackendOptions: NewBackendOptions(),
+	}
+
+	for _, opt := range opts {
+		opt(backend)
+	}
+
+	backend.init(backend.BackendOptions)
+
+	return backend
+}
+
+func (backend *Backend) init(opts *BackendOptions) {
+	if backend.BackendOptions == nil {
+		backend.BackendOptions = NewBackendOptions()
+	}
+
+	// Register the SQLite driver as "sqlite3".
+	if !slices.Contains(sql.Drivers(), "sqlite3") {
+		sqlite.RegisterAsSQLITE3()
+	}
+
+	clientOpts := []ent.Option{}
+	if opts.debug {
+		clientOpts = append(clientOpts, ent.Debug())
+	}
+
+	client, err := ent.Open("sqlite3", opts.DatabaseFile+dsnParams, clientOpts...)
+	if err != nil {
+		panic(fmt.Errorf("failed opening connection to sqlite: %w", err))
+	}
+
+	opts.client = client
+	opts.ctx = ent.NewContext(context.Background(), client)
+
+	// Run the auto migration tool.
+	if err := opts.client.Schema.Create(opts.ctx); err != nil {
+		panic(fmt.Errorf("failed creating schema resources: %w", err))
+	}
+}
+
+func (backend *Backend) Debug() *Backend {
+	backend.debug = true
+	backend.init(backend.BackendOptions)
```

Comment:
Can this function just set the backend option and modify the setting in the client? I feel re-running `init()` could cause some unexpected consequences in the future if the idempotency guarantees are not kept, and since the ent backend uses third party plugins I feel it's just a question of time until it happens.

Reply:
I believe my thought was that `backend.client` might be nil here. Would you rather something like

```go
if backend.Options.client != nil {
	backend.Options.client.Debug()
}
```

That `backend.Options.client.Debug()` method is from the generated ent code, it looks like it creates and returns a new client no matter what. Definitely open to ideas, the only other things I can think of are to either

- drop these "builder" methods and require all options are passed in to `NewBackend()` (not a big fan of this personally), or
- figure out a way to get by without calling `backend.init()` within any of the Store/Retrieve methods

Reply:
See https://github.com/protobom/storage/pull/4#issuecomment-2147895186 for how this was reworked